prov/util/src/util_ns.c

```
@@ -0,0 +1,505 @@
+			      void **name_out, size_t *name_len)
+{
+	RbtIterator it;
+	void *key;
+	struct util_ns_entry *entry;
+
+        it = rbtFind(ns->ns_map, (void *)(uintptr_t)(*service));
+	if (!it)
+		return -FI_ENOENT;
+
+	rbtKeyValue(ns->ns_map, it, &key, (void **)&entry);
+	*name_out = entry->buf;
+	*name_len = entry->len;
+
+	if (*service == 0)
+		*service = (uintptr_t)key;
+
+	return FI_SUCCESS;
+}
+
+static void util_ns_name_server_cleanup(void *args)
+{
+	void **cleanup_args = (void **)args;
+	ofi_close_socket((uintptr_t)cleanup_args[0]);
+	util_ns_map_fini((struct util_ns *)cleanup_args[1]);
+}
+
+static int util_ns_op_dispatcher(struct util_ns *ns,
+				 struct util_ns_cmd *cmd,
+				 SOCKET sock)
+{
+	int ret;
+	void *name_buf = NULL;
+
+	switch (cmd->op) {
+		case OFI_UTIL_NS_ADD:
+			name_buf = calloc(cmd->name_len, 1);
+			if (!name_buf)
+				return -FI_ENOMEM;
+			ret = ofi_read_socket(sock, name_buf,
+					      cmd->name_len);
+			if (ret == cmd->name_len) {
+				ret = util_ns_map_add(
+					ns, cmd->service,
+					name_buf, cmd->name_len
+				);
+				if (ret)
+					free(name_buf);
+			} else {
+				free(name_buf);
+				ret = -FI_ENODATA;
+			}
+			break;
+
+		case OFI_UTIL_NS_DEL:
+			name_buf = calloc(cmd->name_len, 1);
+			if (!name_buf)
+				return -FI_ENOMEM;
+			ret = ofi_read_socket(sock, name_buf,
+					      cmd->name_len);
+			if (ret == cmd->name_len) {
+				ret = util_ns_map_del(
+					ns, cmd->service,
+					name_buf, cmd->name_len
+				);
+			} else {
+				ret = -FI_ENODATA;
+			}
+			free(name_buf);
+			break;
+
+		case OFI_UTIL_NS_QUERY:
+			cmd->op = OFI_UTIL_NS_ACK;
+			cmd->status = util_ns_map_lookup(
+				ns, &cmd->service,
+				&name_buf, &cmd->name_len
+			);
+			ret = ofi_write_socket(sock, cmd,
+					       sizeof(*cmd));
+			if (ret == sizeof(*cmd)) {
+				ret = ofi_write_socket(sock, name_buf,
+						       cmd->name_len);
+				if (ret == sizeof(cmd->name_len)) {
+					ret = FI_SUCCESS;
+				}
+			}
```

If either of the above socket writes fail, we should return a failure code.